runtime/parachains/src/paras.rs

```diff
@@ -892,55 +1422,156 @@ impl<T: Config> Pallet<T> {
+	/// the relay-chain block number will be determined at which the upgrade will take place. We
+	/// call that block `expected_at`.
 	///
+	/// Once the candidate with the relay-parent >= `expected_at` is enacted, the new validation code
+	/// will be applied. Therefore, the new code will be used to validate the next candidate.
+	///
+	/// The new code should not be equal to the current one, otherwise the upgrade will be aborted.
 	/// If there is already a scheduled code upgrade for the para, this is a no-op.
 	pub(crate) fn schedule_code_upgrade(
 		id: ParaId,
 		new_code: ValidationCode,
 		relay_parent_number: T::BlockNumber,
 		cfg: &configuration::HostConfiguration<T::BlockNumber>,
 	) -> Weight {
-		<Self as Store>::FutureCodeUpgrades::mutate(&id, |up| {
-			if up.is_some() {
-				T::DbWeight::get().reads_writes(1, 0)
-			} else {
-				let expected_at = relay_parent_number + cfg.validation_upgrade_delay;
-				let next_possible_upgrade_at =
-					relay_parent_number + cfg.validation_upgrade_frequency;
+		let mut weight = T::DbWeight::get().reads(1);
 
-				*up = Some(expected_at);
+		// Enacting this should be prevented by the `can_schedule_upgrade`
+		if FutureCodeHash::<T>::contains_key(&id) {
+			// This branch should never be reached. Signalling an upgrade is disallowed for a para
+			// that already has one upgrade scheduled.
+			//
+			// Any candidate that attempts to do that should be rejected by
+			// `can_upgrade_validation_code`.
+			UpgradeGoAheadSignal::<T>::insert(&id, UpgradeGoAhead::Abort);
+			log::warn!(
+				target: "runtime::paras",
+				"ended up scheduling an upgrade while one is pending",
+			);
+			return weight
+		}
 
-				<Self as Store>::UpcomingUpgrades::mutate(|upcoming_upgrades| {
-					let insert_idx = upcoming_upgrades
-						.binary_search_by_key(&expected_at, |&(_, b)| b)
-						.unwrap_or_else(|idx| idx);
-					upcoming_upgrades.insert(insert_idx, (id, expected_at));
-				});
+		let code_hash = new_code.hash();
 
-				// From the moment of signalling of the upgrade until the cooldown expires, the
-				// parachain is disallowed to make further upgrades. Therefore set the upgrade
-				// permission signal to disallowed and activate the cooldown timer.
-				<Self as Store>::UpgradeRestrictionSignal::insert(&id, UpgradeRestriction::Present);
-				<Self as Store>::UpgradeCooldowns::mutate(|upgrade_cooldowns| {
-					let insert_idx = upgrade_cooldowns
-						.binary_search_by_key(&next_possible_upgrade_at, |&(_, b)| b)
-						.unwrap_or_else(|idx| idx);
-					upgrade_cooldowns.insert(insert_idx, (id, next_possible_upgrade_at));
-				});
+		// para signals an update to the same code? This does not make a lot of sense, so abort the
+		// process right away.
+		weight += T::DbWeight::get().reads(1);
+		if CurrentCodeHash::<T>::get(&id) == Some(code_hash) {
+			UpgradeGoAheadSignal::<T>::insert(&id, UpgradeGoAhead::Abort);
```

Comment:
nit: Adding a log message here would be appreciated.

Reply:
Good call

Reply:
will be fixed in a follow-up of a follow-up